local-volume/provisioner/pkg/discovery/discovery.go

```diff
@@ -80,49 +119,36 @@ func (d *Discoverer) validateFile(fullPath string) error {
 	return nil
 }
 
-// TODO: maybe a better way would be to hash the 3 fields
 func generatePVName(file, node, class string) string {
-	return fmt.Sprintf("%v-%v-%v", class, node, file)
+	h := fnv.New32a()
+	h.Write([]byte(file))
+	h.Write([]byte(node))
+	h.Write([]byte(class))
+	// This is the FNV-1a 32-bit hash
+	return fmt.Sprintf("local-pv-%x", h.Sum32())
 }
 
 func (d *Discoverer) createPV(file, relativePath, class string) {
-	pvName := generatePVName(file, d.NodeName, class)
+	pvName := generatePVName(file, d.Node.Name, class)
 	outsidePath := filepath.Join(d.HostDir, relativePath, file)
 
 	glog.Infof("Found new volume at host path %q, creating Local PV %q", outsidePath, pvName)
-	pvSpec := &v1.PersistentVolume{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: pvName,
-			Annotations: map[string]string{
-				types.AnnProvisionedBy: d.Name,
-				// TODO: add topology constraint once we have API
-			},
-		},
-		Spec: v1.PersistentVolumeSpec{
-			PersistentVolumeReclaimPolicy: v1.PersistentVolumeReclaimDelete,
-			Capacity: v1.ResourceList{
-				// TODO: detect capacity
-				v1.ResourceName(v1.ResourceStorage): resource.MustParse("10Gi"),
-			},
-			PersistentVolumeSource: v1.PersistentVolumeSource{
-			/** TODO: api not merged yet
-			LocalVolume: &v1.LocalVolumeSource{
-				NodeName: config.NodeName,
-				Fs: v1.LocalFsVolume{Path: outsidePath},
-			},*/
-			},
-			AccessModes: []v1.PersistentVolumeAccessMode{
-				v1.ReadWriteOnce,
-			},
-			StorageClassName: class,
-		},
-	}
-
-	pv, err := d.APIUtil.CreatePV(pvSpec)
+	pvSpec := common.CreateLocalPVSpec(&common.LocalPVConfig{
+		Name:            pvName,
+		HostPath:        outsidePath,
+		StorageClass:    class,
+		ProvisionerName: d.Name,
+		AffinityAnn:     d.nodeAffinityAnn,
+	})
+
+	// Add to cache first to handle race condition between discoverer and informer updating the same pv in the cache
+	// This way the ordering is consistent. First discoverer adds to cache, then informer will update the pv object
+	d.Cache.AddPV(pvSpec)
```

Comment:
Yeah, the idea was to immediately add it to the cache after creating, so that we don't try to create the PV twice.  But I suppose there is no harm in trying to create twice, since we don't do any sort of formatting or destructive action here.  I can remove the cache add here and just let the informer do it.